### PR TITLE
update(user): Integrate validation in user creation endpoint (#650)

### PR DESCRIPTION
**Pull request by isaac-diez**
### Summary

This PR updated the behaviour of AdminCreateUserService() so the username is checked against the  a centralized Github Username Validation that will be shared with User and Auth microservices. 

This PR is the second part of 3 which are derived from the following tasks:
- [#649](https://tree.taiga.io/project/luisvi-ita-challenges/task/649) 1/3 Create GitHub Username Validation in Shared Module
- [#650](https://tree.taiga.io/project/luisvi-ita-challenges/task/650) 2/3 Integrate Validation in User Creation Endpoint
- [#651](https://tree.taiga.io/project/luisvi-ita-challenges/task/651) 3/3 Refactor & Reuse in Auth-Service

**Context**

When creating a new user in the user microservice, the GitHub username provided needed to be validated against GitHub’s user base to ensure it actually exists. Previously, this validation was missing.

**Before**

The user microservice allowed creating users with any GitHub username, even if it did not exist. No external check against GitHub’s API was performed.

**After**

The user microservice now validates GitHub usernames by calling the GitHub API via the new github-core module. 
On user creation, the service checks if the GitHub user exists and rejects invalid usernames.

**Impact**

- Breaking change: Clients must now provide valid GitHub usernames when creating users (invalid ones will be rejected with a validation error).
- Functional case (GithubUserStatus.NOT_FOUND) is now translated in user-service into a NotFoundException, which results in 404 Not Found to clients.
- Provider failures (GithubUnavailableException, timeouts, 5xx) are surfaced from github-core and consistently mapped in GlobalExceptionHandler to 503 Service Unavailable/504 Gateway Timeout.

IMPORTANT: 
-
**Since this task depends on the #649 all its changes and new contributions also appear in this PR, so refrain from checking everything that falls under the github-core folder.**

### Main tasks:
- Refactor AdminCreateUserService to implement Github username validation for which the creation of a new user is depends on the username being a Github user.
- Update all dependent classes.
- Keep all tests green in case the new implementation needs some refactoring.

### Changes 

**App.java**
Added annotation to import the GithubServiceConfig() class to get the Github API URL to validate the username.
It is important because, this class needs the github user info uri to be in the application.yml or as environment variables of every micro that is using the Github-core module until it is moved into a central config server. Here is, for reference, the GithubServiceConfig():

    @Configuration
    public class GithubServiceConfig {

    @Value("${github.user-info-uri}")
    private String githubApiUrl;

    @Bean
    public GithubApiService githubApiService(WebClient.Builder webClientBuilder) {
        return new GithubApiServiceImpl(webClientBuilder, githubApiUrl);
        }
    }

and the application.yml:

    github:
      user-info-uri: https://api.github.com

**ExternalGithubService  & ExternalGithubServiceImpl**
In order to fully decouple the main service from the github-core module, this interface and its implementation take care of passing true/false on github username validation to the main service.

**AdminCreateUserService**
Passed the interface ExternalGithubService as an argument and run the username by its userExists() method right after checking if the user is alredy in the database.

**APIErrorResponse**
Created to conveniently inform when the Github API fails.
--> TECHNICAL DEBT: Right now the handler is misaligned so the APIResponse should be adapted  to the rest of the exceptions so that they follow the same format.

**UserGlobalExceptionHandler**
Added the GithubUnavailableException handler for when the Github API is unavailable (code 503) or times out (504), returning an APIErrorResponse DTO.
--> TECHNICAL DEBT: a more robust solution to throw codes 503 or 504 and informs more securely.

**AdminCreateUserServiceTest**
Added the following tests:
- for the situation when the user in not on the database but neither is a Github user.
- for when the Github API fails and should propagate error
- to check that the new user should always have default USER role

**ExternalGithubServiceTest & APIErrorResponseTest**
Created to test the new middleware service and DTO.

**AdminCreateUserControllerTest**
The tests were not returning the right server codes because they needed injecting of @ContextConfiguration and @Import of the UserGlobalExceptionHandler.class to throw the right exceptions and Http responses. 

**UserGlobalExceptionHandlerTest**
Tested the new GithybUnavailableException handler that checks whether a 503 or 504 code is thrown.

**UserControllerSpringTest & UserSolucitonServiceImplTest**
The new ExternalGithubService bean is needed to be injected for the test to work.

**settings.gradle**
Include github-core for the build to use the new module and its services.

**CHANGELOG.md & .env.CI,dev**
Moved to version 1.5.0 of user microservice.

**README.md** for Github-core module
This was left from the previous PR where i created the module. It explains what the new module Github-core does, structure and usage.

**POSTMAN TESTS**
Screenshot-2025-09-16-17:38:27

## 📋 PR Author Self-check
✅ PR is focused and isolated
✅ Branch follows naming conventions.
✅ No debug logs left.
✅ Exception mapped correctly.
✅ PR clearly documents scope and purpose.

### CODE REVIEW

**Review by acocinas — APPROVED** [review]
Hey Isaac,

Great work on this PR! The implementation is solid and the integration with the github-core module is very clean. The test coverage is excellent, and it's clear you've considered all the main scenarios.

The code is in a good state and ready to be merged. I'm leaving a minor comment on the CHANGELOG for clarification, but it doesn't block the PR.


Great job!

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,11 @@
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+### [itachallenge-user-1.5.0-RELEASE] - 2025-09-12
+
+### Refactored
```
Comment (acocinas) [inline CHANGELOG.md]: Hi. Just a quick question on the CHANGELOG: would ### Added or ### Changed be a more accurate label for this entry? The task implemented a new validation feature rather than just an internal refactor. Just a thought for consistency.

**Review by Heyvkram — APPROVED** [review]
Echoing the previous comment, this is a fantastic PR, Isaac. The implementation is very solid, and the decision to create a separate github-core module is a great architectural choice that's been cleanly integrated.

The test coverage is comprehensive, and it's clear all edge cases were considered. 

Take a look at the comment @acocinas made about the CHANGELOG!

Excellent work.

File: CHANGELOG.md (hunk 1 above)
Comment (isaac-diez) [reply CHANGELOG.md]: OK, I see that now. Yes, it is more accurate. I change that.

Comment (acocinas) [reply CHANGELOG.md]: ok buddy, thank you

**Comment by ivilarop** [comment]
> ### Summary
> This PR updated the behaviour of AdminCreateUserService() so the username is checked against the a centralized Github Username Validation that will be shared with User and Auth microservices.
> 
> This PR is the second part of 3 which are derived from the following tasks:
> 
> * [#649](https://tree.taiga.io/project/luisvi-ita-challenges/task/649) 1/3 Create GitHub Username Validation in Shared Module
> * [#650](https://tree.taiga.io/project/luisvi-ita-challenges/task/650) 2/3 Integrate Validation in User Creation Endpoint
> * [#651](https://tree.taiga.io/project/luisvi-ita-challenges/task/651) 3/3 Refactor & Reuse in Auth-Service
> 
> ## IMPORTANT:
> **Since this task depends on the #649 all its changes and new contributions also appear in this PR, so refrain from checking everything that falls under the github-core folder.**
> 
> ### Main tasks:
> * Refactor AdminCreateUserService to implement Github username validation for which the creation of a new user is depends on the username being a Github user.
> * Update all dependent classes: App.java and application.yml
> * Keep all tests green in case the new implementation needs some refactoring.
> 
> ### Changes
> **App.java** Added annotation to use the GithubServiceConfig class to get the Github API URL to validate the username
> 
> **AdminCreateUserService** Passed the interface IGithubApiService as an argument and run the username by its userExists() method right after checking if the user is alredy in the database.
> 
> **UserGlobalExceptionHandler** Added the GithubUserNotFoundException handler for when the username is not a Github user the exception is thrown.
> 
> **AdminCreateUserServiceTest** Added the following tests:
> 
> * for the situation when the user in not on the database but neither is a Github user.
> * for when the Github API fails and should propagate error
> * to check that the new user should always have default USER role
> 
> **AdminCreateUserControllerTest** The tests were not returning the right server codes because they needed injecting of @ContextConfiguration and @import of the UserGlobalExceptionHandler.class to throw the right exceptions and Http responses.
> 
> **settings.gradle** Include github-core for the build to use the new module and its services.
> 
> **IGithubApiService & GithubApiService** Interface and implementation that take care of the mail logic of validating the username with Github. Returns true if server returns a 200 cods or false if the server returns a 4xx or 5xx code.
> 
> **CHANGELOG.md & .env.CI,dev** Moved to version 1.5.0 of user microservice.
> 
> **README.md** Explaining what this module does, structure and usage.
> 
> **POSTMAN TESTS** Screenshot-2025-09-16-17:38:27
> 
> ## 📋 PR Author Self-check
> ✅ PR is focused and isolated ✅ Branch follows naming conventions. ✅ No debug logs left. ✅ Exception mapped correctly. ✅ PR clearly documents scope and purpose.

@isaac-diez @Heyvkram 

One thing you could reflect on: the PR description is very detailed about the what (classes, methods, validation), but it’s missing some higher-level context.

Ask yourself:

1. Which microservice does this PR affect?
2. How does it fit into the current login flow?
3. What was the behavior before? What is the behavior after?
4. What is the impact for API consumers (SemVer)?

Adding a short “Context / Before / After / Impact” section would make the PR description easier to understand for reviewers who are not deep into the code.

What do you think — could you try to rewrite the PR description with this extra context?

File: CHANGELOG.md [hunk 2]
```diff
@@ -3,6 +3,11 @@
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+### [itachallenge-user-1.5.0-RELEASE] - 2025-09-12
```
Comment (ivilarop) [inline CHANGELOG.md]: One thing you should reflect on is the versioning choice.
Ask yourself:

1. Before this change, what requests did the user microservice accept?
2. After this change, what requests are now rejected?
3. Does this break compatibility for any existing clients?
4. Based on Semantic Versioning, what kind of release would that imply?

Try to reason it out and update the changelog accordingly.

File: conf/.env.CI.dev [hunk 3]
```diff
@@ -2,7 +2,7 @@ MICROSERVICE_DEPLOY_itachallenge-auth=itachallenge-auth
 MICROSERVICE_VERSION_itachallenge-auth=2.0.5-RELEASE
 
 MICROSERVICE_DEPLOY_itachallenge-user=itachallenge-user
-MICROSERVICE_VERSION_itachallenge-user=1.4.2-RELEASE
+MICROSERVICE_VERSION_itachallenge-user=1.5.0-RELEASE
```
Comment (ivilarop) [inline conf/.env.CI.dev]: see previous comment

File: github-core/README.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
+# ITA GITHUB-CORE MODULE
```
Comment (ivilarop) [inline github-core/README.md:1]: Why did this change happen? Does it have to do with the PR? I know you said, "Since this task depends on #649, all its changes and new contributions also appear in this PR, so refrain from checking everything that falls under the github-core folder." But I don't see this change in #649.
You also added a description of this change to the PR.

File: itachallenge-user/src/main/java/com/itachallenge/user/service/AdminCreateUserService.java [hunk 5]
```diff
@@ -18,9 +21,11 @@ public class AdminCreateUserService implements IAdminCreateUserService {
     private static final Logger log = LoggerFactory.getLogger(AdminCreateUserService.class);
 
     private final UserRepository userRepository;
+    private final GithubApiService githubApiService;
```
Comment (ivilarop) [inline itachallenge-user/src/main/java/com/itachallenge/user/service/AdminCreateUserService.java]: But take a moment to think about this:

1. How coupled is the user microservice to the details of github-core right now?
2. What would break if github-core changes its API or internals?
3. Is this the same level of isolation you achieved when integrating jwt-core?

Please review with these questions in mind and adjust so the micro is less dependent on the internals of github-core.

File: itachallenge-user/src/main/resources/application.yml [hunk 6]
```diff
@@ -32,6 +32,9 @@ server:
   tomcat:
     max-http-form-post-size: 2097152
 
+github:
```
Comment (ivilarop) [inline itachallenge-user/src/main/resources/application.yml:35]: The GitHub base URL shouldn’t live in the user micro’s application.yml.
That config belongs to github-core

File: itachallenge-user/src/main/java/com/itachallenge/user/service/AdminCreateUserService.java [hunk 7]
```diff
@@ -29,34 +34,36 @@ public Mono<AdminCreateUserResponseDto> createUser(AdminCreateUserRequestDto req
...
+                                    return userRepository.save(newUser)
+                                            .map(savedUser -> AdminCreateUserResponseDto.builder()
+                                                    .userId(savedUser.getUuid().toString())
+                                                    .username(savedUser.getUsername())
+                                                    .role(savedUser.getRole().toString())
+                                                    .build())
+                                            .doOnSuccess(responseDto ->
+                                                    log.info("Successfully created user '{}'", responseDto.getUsername()));
+                                })
+                                .doOnError(e -> !(e instanceof UsernameAlreadyExistsException || e instanceof GithubUserNotFoundException), e ->
```
Comment (ivilarop) [inline itachallenge-user/src/main/java/com/itachallenge/user/service/AdminCreateUserService.java]: **Current behavior**

1. In AdminCreateUserService you use doOnError(...) to log unexpected errors.
2. doOnError only performs a side-effect (logging). It does not change the reactive flow and does not translate provider errors into an API response.

**Why this is a problem**

1. When github-core fails (5xx/timeout/unexpected), the service just logs and lets it bubble up, which typically ends as a generic 500.
2. This duplicates logging and hides the intent; API consumers don’t get a clear, consistent status (e.g., 503).

**What we should do instead**

**1.** Remove the doOnError used to “handle” provider failures (it doesn’t handle them).
**2.** Handle outcomes centrally in the @ControllerAdvice:

- Provider failures: do not create a new exception in user — import com.itachallenge.githubcore.exception.GithubUnavailableException (from github-core) in the handler and return 503 Service Unavailable (or 504 on timeouts).
- Functional “user not found”: keep it as normal flow from github-core (FOUND/NOT_FOUND).
 In the user micro, translate NOT_FOUND to a domain not-found exception (you can reuse your existing NotFoundException) so the handler returns 404.

**3.** If you need to transform exceptions in the service layer, use onErrorMap/onErrorResume (which affect the flow), not doOnError.
**4.** (Nice to have) Return a structured JSON error payload instead of raw strings for consistency.

Please update the error handling accordingly and add a short note in the PR description explaining:

- how provider failures are mapped (503/504) and
- how the NOT_FOUND case is translated to 404.

File: itachallenge-user/src/main/java/com/itachallenge/user/exception/GithubUserNotFoundException.java
Comment (ivilarop) [inline itachallenge-user/src/main/java/com/itachallenge/user/exception/GithubUserNotFoundException.java:1]: Don't create a new exception. Reuse the one you already have:

When github-core returns NOT_FOUND (enum/empty), throw new NotFoundException("GitHub user not found: " + username) in your service.

You already have a NotFoundException handler → it responds with a 404. ✔️

File: itachallenge-user/src/main/java/com/itachallenge/user/exception/UserGlobalExceptionHandler.java [hunk 8]
```diff
@@ -82,4 +82,9 @@ public ResponseEntity<String> handleUsernameAlreadyExistsException(UsernameAlrea
         return ResponseEntity.status(HttpStatus.CONFLICT).body(e.getMessage());
     }
 
+    @ExceptionHandler(GithubUserNotFoundException.class)
```
Comment (ivilarop) [inline itachallenge-user/src/main/java/com/itachallenge/user/exception/UserGlobalExceptionHandler.java]: see previous comment

**Review by ivilarop — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 2 above)
Comment (isaac-diez) [reply CHANGELOG.md]: After some analysis, the changes affect the API contract  because some previously valid requests will now be rejected. According to SemVer that is a breaking change so a major version bump is appropriate. I, then, moved the version to 2.0.

File: conf/.env.CI.dev (hunk 3 above)
Comment (isaac-diez) [reply conf/.env.CI.dev]: As per my previous comment's response, I have moved the version to 2.0.0

File: github-core/README.md (hunk 4 above)
Comment (isaac-diez) [reply github-core/README.md:1]: I wanted to push as quickly as possible task #649 so I decided to postpone the Github-core README document to this task's (#650) PR.

File: itachallenge-user/src/main/java/com/itachallenge/user/exception/GithubUserNotFoundException.java
Comment (isaac-diez) [reply itachallenge-user/src/main/java/com/itachallenge/user/exception/GithubUserNotFoundException.java:1]: I created a new exception with clarity and single responsibility in mind. But it might reduce maintainability.
I have deleted the GithubUserNotFoundException and used NotFoundException instead.

File: itachallenge-user/src/main/java/com/itachallenge/user/exception/UserGlobalExceptionHandler.java (hunk 8 above)
Comment (isaac-diez) [reply itachallenge-user/src/main/java/com/itachallenge/user/exception/UserGlobalExceptionHandler.java]: It's gone now.

File: itachallenge-user/src/main/java/com/itachallenge/user/service/AdminCreateUserService.java (hunk 5 above)
Comment (isaac-diez) [reply itachallenge-user/src/main/java/com/itachallenge/user/service/AdminCreateUserService.java]: After deeper analysis, I can see that the service is highly coupled with the new Github-core module because it directly depends on:

- GithubApiService (the service interface/class from github-core).
- GithubUserStatus (an enum defined inside github-core).

The user micro is making branching logic (if (status == GithubUserStatus.NOT_FOUND)) based on an internal detail of how github-core treats GitHub responses.

This means if github-core changes its enum values, or switches to a different mechanism (e.g., returning a DTO or throwing exceptions instead of GithubUserStatus), the user micro will break.

For this, i have refactored the service into using an abstraction layer with a new ExternalGithubService that takes care of the boolean response of userExists(username) so the main services only depends on a boolean answer and is decoupled from the new Github-core module.

File: itachallenge-user/src/main/java/com/itachallenge/user/service/AdminCreateUserService.java (hunk 7 above)
Comment (isaac-diez) [reply itachallenge-user/src/main/java/com/itachallenge/user/service/AdminCreateUserService.java]: You are right. I was only throwing the log.error without specific exception. I have refactored the service to throw GithubUnavailableException to return 503 or 504 when needed.

Functional case (GithubUserStatus.NOT_FOUND) is now translated in user-service into a NotFoundException, which results in 404 Not Found to clients.

Provider failures (GithubUnavailableException, timeouts, 5xx) are surfaced from github-core and consistently mapped in GlobalExceptionHandler to 503 Service Unavailable/504 Gateway Timeout.

Removed doOnError in AdminCreateUserService to avoid duplicated logging and misleading error handling.

All error responses now follow a structured JSON format for consistency with a newly created APIErrorResponse DTO.

File: itachallenge-user/src/main/resources/application.yml (hunk 6 above)
Comment (isaac-diez) [reply itachallenge-user/src/main/resources/application.yml:35]: Since github-core is a library and not a Spring Boot app, it cannot own application.yml (Spring won’t even load it). The correct approach is:

- github-core defines what configuration keys it expects (github.user-info-uri).

- Each runnable microservice (user-micro, auth-micro, …) provides values in its application.yml (or better, in Spring Cloud Config if you want centralization).

So unless everything is moved to a central config server, the application.yml entry must stay in the runnable microservice(s).

**Review by acocinas — APPROVED** [review]
I have reviewed all the changes and the evolution of this PR. The work has done is, quite simply, outstanding. You not only solved the initial task, but also took very technical and complex feedback and implemented it flawlessly, significantly improving the microservice's architecture.

Final Analysis
What started as a good PR for validation has evolved into an exceptional piece of code thanks to the review process. Significant improvements have been made at every layer:

Reduced Coupling: The creation of the ExternalGithubService interface is a brilliant solution. It completely decouples the user microservice from the internal details of github-core, such as the GithubUserStatus enum. The user service now only depends on a simple Mono<Boolean>, which makes the code far more flexible and maintainable.

Robust Error Handling: The centralization of error handling in UserGlobalExceptionHandler, the use of onErrorMap, and the structured APIErrorResponse demonstrate an advanced understanding of Spring Boot and reactive best practices.

Adherence to Standards: The versioning correction to 2.0.0 and the justification for the breaking change are a clear example of a mature approach to Semantic Versioning.

This PR has gone through a rigorous review process and has been greatly improved as a result.

**Review by Heyvkram — APPROVED** [review]
Approved.

This is an excellent contribution. The preceding review has already highlighted the significant technical achievements in this PR, and I am in full agreement. The evolution of the code through the review process has resulted in a robust and maintainable solution. 
Good job!

File: CHANGELOG.md (hunk 2 above)
Comment (ivilarop) [reply CHANGELOG.md]: Ok perfect! its ok!

File: conf/.env.CI.dev (hunk 3 above)
Comment (ivilarop) [reply conf/.env.CI.dev]: Ok perfect! its ok!

File: github-core/README.md (hunk 4 above)
Comment (ivilarop) [reply github-core/README.md:1]: If you include a change that might seem out of scope (such as the GitHub-core README), it's important to also make it explicit in the PR description. This way, any reviewer immediately understands why it's there, and clear traceability between tasks and PRs is maintained.

Please leave it noted in the PR description.

File: itachallenge-user/src/main/java/com/itachallenge/user/service/AdminCreateUserService.java (hunk 5 above)
Comment (ivilarop) [reply itachallenge-user/src/main/java/com/itachallenge/user/service/AdminCreateUserService.java]: Great reasoning and solid refactor to reduce coupling 👏.
One minor suggestion for consistency: in the auth micro we talk to jwt-core through a facade (IAuthJwtFacade / AuthJwtFacade).
To keep naming coherent across services, could we mirror that here for github-core?

This way any dev recognizes the pattern immediately and we avoid leaking provider details into the domain.

File: itachallenge-user/src/main/resources/application.yml (hunk 6 above)
Comment (ivilarop) [reply itachallenge-user/src/main/resources/application.yml:35]: Ok, I understand now: the configuration value (e.g. github.user-info-uri) must indeed live in the user micro (or any other runnable microservice) and then be provided to the github-core library at runtime. That part makes sense 👍.

What I’m missing, however, is that this is not explained anywhere in the PR description. Right now it’s not clear for another developer why the property has to be in the micro’s application.yml instead of inside github-core.

I suggest adding a short Configuration section in the PR description to clarify:

- github-core defines the property key(s) it expects.
- Each runnable microservice must provide the value in its own application.yml or via env vars.
- Example snippet of how to set it up.

This way, anyone reading the PR understands both the technical change and the configuration requirement.

File: itachallenge-user/src/main/java/com/itachallenge/user/service/AdminCreateUserService.java (hunk 7 above)
Comment (ivilarop) [reply itachallenge-user/src/main/java/com/itachallenge/user/service/AdminCreateUserService.java]: OK! The current Boolean mapping (status != NOT_FOUND) is fine as an MVP ✅, but it hides some domain intent.
As technical debt, it would be worth replacing the Boolean with a more explicit result type in the future. That way the meaning is clearer (e.g. FOUND vs NOT_FOUND) and the compiler will force us to handle any new cases if github-core evolves.

If a new value (e.g., RATE_LIMITED) appears in the future, the compiler will force the service to decide how to handle it. It won't accidentally "sneak" in as true.

Not blocking for now, but good to keep in mind for a follow-up improvement.

File: itachallenge-user/src/main/java/com/itachallenge/user/exception/GithubUserNotFoundException.java
Comment (ivilarop) [reply itachallenge-user/src/main/java/com/itachallenge/user/exception/GithubUserNotFoundException.java:1]: Ok thks! its ok!

File: itachallenge-user/src/main/java/com/itachallenge/user/exception/UserGlobalExceptionHandler.java (hunk 8 above)
Comment (ivilarop) [reply itachallenge-user/src/main/java/com/itachallenge/user/exception/UserGlobalExceptionHandler.java]: Ok thks! its ok!

File: itachallenge-user/src/main/java/com/itachallenge/user/dto/APIErrorResponse.java
Comment (ivilarop) [inline itachallenge-user/src/main/java/com/itachallenge/user/dto/APIErrorResponse.java:1]: It's good that it was created, and that's what I told you, but one clarification: Currently, it's only being used in the GithubUnavailableException handler.

It would be good to extend it to all handlers (404, 409, etc.) so that the client always receives the same format.

APIErrorResponse uses error + message, but there's no stable code/key, status, path, traceId, etc. It's difficult to parse from the front end and correlate in logs. (Optional: browse Problem Details: type, title, status, detail, instance.)

Have you considered this as technical debt?

Please note that we should modify the PR description to consider that right now the handler is misaligned and that we are going to create a technical debt to adapt the APIResponse to the rest of the exceptions so that they follow the same format.

File: itachallenge-user/src/main/java/com/itachallenge/user/exception/UserGlobalExceptionHandler.java [hunk 9]
```diff
@@ -82,4 +85,19 @@ public ResponseEntity<String> handleUsernameAlreadyExistsException(UsernameAlrea
         return ResponseEntity.status(HttpStatus.CONFLICT).body(e.getMessage());
     }
 
+    @ExceptionHandler(GithubUnavailableException.class)
+    public ResponseEntity<APIErrorResponse> handleGithubUnavailable(GithubUnavailableException ex) {
+        HttpStatus status;
+
+        if ("timeout".equalsIgnoreCase(ex.getMessage())) {
```
Comment (ivilarop) [inline itachallenge-user/src/main/java/com/itachallenge/user/exception/UserGlobalExceptionHandler.java:92]: It works, but string matching in getMessage() is fragile.

Deciding 503 vs. 504 with ex.getMessage().equalsIgnoreCase("timeout") is fragile: if the message changes, if someone internationalizes it or adjusts it, you'll inadvertently break the mapping.

File: itachallenge-user/src/main/java/com/itachallenge/user/exception/UserGlobalExceptionHandler.java [hunk 10]
```diff
@@ -82,4 +85,19 @@ public ResponseEntity<String> handleUsernameAlreadyExistsException(UsernameAlrea
...
+            status = HttpStatus.GATEWAY_TIMEOUT; // 504
+        } else {
+            status = HttpStatus.SERVICE_UNAVAILABLE; // 503
+        }
+
+        return ResponseEntity.status(status).body(
+                new APIErrorResponse("GitHub API error", ex.getMessage(), Instant.now())
```
Comment (ivilarop) [inline itachallenge-user/src/main/java/com/itachallenge/user/exception/UserGlobalExceptionHandler.java:99]: ex.getMessage() goes to the client as is. It may leak internal details or vary between environments. A "secure" message and detailed internal log are preferred.

**Review by ivilarop — CHANGES_REQUESTED** [review]

File: github-core/README.md (hunk 4 above)
Comment (isaac-diez) [reply github-core/README.md:1]: Ok, noted. The summary is updated now.

File: itachallenge-user/src/main/java/com/itachallenge/user/dto/APIErrorResponse.java
Comment (isaac-diez) [reply itachallenge-user/src/main/java/com/itachallenge/user/dto/APIErrorResponse.java:1]: Yes, I see your point and it makes total sense. I'll note it as technical debt.

File: itachallenge-user/src/main/java/com/itachallenge/user/exception/UserGlobalExceptionHandler.java (hunk 10 above)
Comment (isaac-diez) [reply itachallenge-user/src/main/java/com/itachallenge/user/exception/UserGlobalExceptionHandler.java:99]: Noted as technical debt as well.

File: itachallenge-user/src/main/java/com/itachallenge/user/exception/UserGlobalExceptionHandler.java (hunk 9 above)
Comment (isaac-diez) [reply itachallenge-user/src/main/java/com/itachallenge/user/exception/UserGlobalExceptionHandler.java:92]: I think this is best left as technical debt. I have checked and it seems that I might be needing to modify the constructor of the exception created in the Github-core module GithubUnavailableException() in order to pass a cause.

**Review by ivilarop — APPROVED** [review]
